catalyst-gateway/bin/src/db/index/index_cip36.rs

```diff
@@ -0,0 +1,243 @@
+//! Index CIP-36 Registrations.
+
+use std::sync::Arc;
+
+use cardano_chain_follower::{Metadata, MultiEraBlock};
+use scylla::{frame::value::MaybeUnset, SerializeRow, Session};
+use tracing::error;
+
+use super::{
+    queries::{FallibleQueryTasks, PreparedQueries, PreparedQuery, SizedBatch},
+    session::CassandraSession,
+};
+use crate::settings::CassandraEnvVars;
+
+/// Insert CIP-36 Registration Query Parameters
+#[derive(SerializeRow, Clone)]
+pub(crate) struct Cip36RegistrationInsertQuery {
+    /// Stake key hash
+    stake_address: MaybeUnset<Vec<u8>>,
+    /// Stake key hash
+    vote_key: Vec<u8>,
+    /// Slot Number the cert is in.
+    slot_no: num_bigint::BigInt,
+    /// Transaction Index.
+    txn: i16,
+    /// Full Payment Address (not hashed, 32 byte ED25519 Public key).
+    payment_address: MaybeUnset<Vec<u8>>,
+    /// Is the stake address a script or not.
+    is_payable: MaybeUnset<bool>,
+    /// Raw nonce value.
+    raw_nonce: MaybeUnset<num_bigint::BigInt>,
+    /// Nonce value after normalization.
+    nonce: MaybeUnset<num_bigint::BigInt>,
+    /// Purpose. Always `0` for Catalyst.
+    purpose: MaybeUnset<num_bigint::BigInt>,
+    /// Signature validates.
+    signed: MaybeUnset<bool>,
+    /// Strict Catalyst validated.
+    strict_catalyst: MaybeUnset<bool>,
+    /// List of serialization errors.
+    error_report: MaybeUnset<Vec<String>>,
+}
+
+/// Index Registration by Stake Address
+const INSERT_CIP36_REGISTRATION_QUERY: &str =
+    include_str!("./queries/insert_cip36_registration.cql");
+
+/// Index Registration by Vote Key
+const INSERT_CIP36_REGISTRATION_BY_VOTE_KEY_QUERY: &str =
+    include_str!("./queries/insert_cip36_vote_key.cql");
+
+impl Cip36RegistrationInsertQuery {
+    /// Create a new Insert Query.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
```

Comment:
what do you think to transfer such constructor function into the `Cip36RegistrationInsertQueryBuilder` struct.
So it will simplify the way of constructing `Cip36RegistrationInsertQuery` as most of the arguments are optional.

It will also allows to change from `Option<Vec<_>>` type to `Vec<_>` and define is it unset or not, checking is the ver is empty or not.

```Rust
struct Cip36RegistrationInsertQueryBuilder {
   stake_address: Vec<u8>,
   payment_address: Vec<u8>,  
    error_report: Vec<String>,
}

impl Cip36RegistrationInsertQueryBuilder {
    fn new(...) -> Self {
      Self {
       ...
      stake_address: Vec::new(),
      payment_address: Vec::new(),
     error_report: Vec::new(),
      }
    }
}
```

Reply:
I've refactored this function, only had to add txn_index and slot_no to the parameters.